crates/sui-types/src/messages.rs

```diff
@@ -95,6 +95,20 @@ pub struct TransferSui {
     pub amount: Option<u64>,
 }
 
+/// Pay each recipient the corresponding amount using the input coins
+#[derive(Debug, PartialEq, Eq, Hash, Clone, Serialize, Deserialize)]
+pub struct Pay {
+    /// The coins to be used for payment
+    pub coins: Vec<ObjectRef>,
```

Comment:
An edge case for this is whether the caller should include the gas object(used to pay for the transaction) in this vector or not if the payment is in SUI. I think the semantics could be if they pass in the gas object id in `coins`(in most cases they should), then it will be used for both gas payment and the payment to recipients. Otherwise, the gas object will only be used for gas payment.

Reply:
`Pay` is a `SingleTransactionKind` which I think should include and only include specs to construct this kind of txn, leaving gas payment to `TransactionData`, as TransactionData` has both `gas_payment: ObjectRef` and `SingleTransactionKind`.
https://github.com/gegaowp/sui/blob/a300658ec5ad1126155982bc5ffad834466a37fc/crates/sui-types/src/messages.rs#L354

I think it is better this way because:
- validity check of `Pay` is easy, we will just make sure sum(coin_balance) >= sum(amounts); otherwise, we will need to consider various gas cost, which is not guaranteed even with gas estimation.
- the problem of "coin selection for payment" is isolated, and its solution can be used for all kinds of txns.

Reply:
Yes, `gas_payment` should always be specified at the top level of `TransactionData`. My comment is more about whether we should allow the same object ref to appear in both `TransactionData.gas_payment` and `Pay.coins`.

Reply:
The intention is that `Pay` works for all coin types (since I think settlement in stablecoins will be common), but reusing the gas coin only applies for payments in SUI.

There *could* be a `PaySui` function that acts like `Pay` but allows you to reuse the gas coin, but I think that this (if it exists at all, which may or may not make sense) should probably be a separate tx type. This way, `Pay` won't have to worry about details like whether coin type is one that allows the gas coin to be debited for payments, whether the gas coin gets debited before or after the rest of the `coins`, and whether an empty list of `coins` is acceptable.

Reply:
If we do want to support a `PaySui` function, I think eliminating `TransferSui` probably makes sense--just generalize it to accepting a (possibly-empty) vector of coins in addition to the gas coin, and a vec of addresses + amounts instead of a single address + amount.